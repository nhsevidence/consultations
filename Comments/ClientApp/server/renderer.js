// See https://medium.com/@cereallarceny/server-side-rendering-with-create-react-app-fiber-react-router-v4-helmet-redux-and-thunk-275cb25ca972
// and https://github.com/cereallarceny/cra-ssr/blob/master/server/universal.js

import { createServerRenderer } from "aspnet-prerendering";
import React from "react";
import { renderToString  } from "react-dom/server";
import { StaticRouter } from "react-router";
import { Helmet } from "react-helmet";
import * as fs from "fs";
import * as path from "path";

import App from "./../src/App";

const OpeningHtmlTagRegex: RegExp = /<html(.*?[^?])?>/,
    OpeningBodyTagRegex: RegExp = /<body(.*?[^?])?>/,
    ClassAttributeRegex: RegExp = /class=\"([^"]*)\"/,
    IsProduction: boolean = process.env.NODE_ENV === "production";

// Replaces the html tag with the given html attributes
const replaceHtmlTag = (html: string, htmlAttributes): string => {
    // Always add a no-js class to support Modernizr
    let className: string = htmlAttributes.toComponent().className || "",
        attrs: string = htmlAttributes.toString().replace(ClassAttributeRegex, "");

    return html.replace(OpeningHtmlTagRegex, `<html class="no-js ${className}" ${attrs}>`);
};

// Replaces the react root div with the given content html
const replaceBodyTag = (html: string, bodyAttributes): string => {
    return html.replace(OpeningBodyTagRegex, `<body ${bodyAttributes}>`);
};

// Replaces the react root div with the given content html
const replaceRootContent = (html: string, rootContent: string): string => {
    return html.replace("<div id=\"root\"></div>", `<div id="root">${rootContent}</div>`);
};

// Replace placeholders and tokens in the static html layout file.
const prepHtml = (html: string, helmet): string => {
    return html
        .replace("<!-- title -->", helmet.title.toString())
        .replace("<!-- meta -->", helmet.meta.toString())
        .replace("<!-- link -->", helmet.link.toString())
        .replace("<!-- script -->", helmet.script.toString());
};

// Returns a promise that resolves to an object containing the HTML to be rendered.
// The params contains properties e.g.
// { location: { protocol, slashes, auth, host, port, hostname, hash, search, query: {}, pathname, path, href },
// origin, url, baseUrl, absoluteUrl, domainTasks: { }, data: { originalHtml: "", ... }
// The `params.data` property contains properties set in `SupplyData` in Startup.cs.
export default createServerRenderer(params => {
    return new Promise((resolve, reject) => {

        // Context object that Routes can use to pass properties 'out'. Primarily used for status code.
        // E.g. 
        //  <Route render={({ staticContext }) => {
        //      if (staticContext) staticContext.status = 404;
        //      return null; }} />
        let context = {};

        const rootContent = renderToString(
            <StaticRouter location={params.url} context={context}>
                <App />
            </StaticRouter>
        );

        const helmet = Helmet.renderStatic();

        let html = params.data.originalHtml;
        html = replaceHtmlTag(html, helmet.htmlAttributes);
        html = replaceBodyTag(html, helmet.bodyAttributes);
        html = replaceRootContent(html, rootContent);
        html = prepHtml(html, helmet);

        resolve({ html: html, statusCode: context.status || 200 });

    });
});