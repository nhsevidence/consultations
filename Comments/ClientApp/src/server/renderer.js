// eslint-disable-next-line
// See https://medium.com/@cereallarceny/server-side-rendering-with-create-react-app-fiber-react-router-v4-helmet-redux-and-thunk-275cb25ca972
// and https://github.com/cereallarceny/cra-ssr/blob/master/server/universal.js

import { createServerRenderer } from "aspnet-prerendering";
import React from "react";
import { renderToString  } from "react-dom/server";
import { StaticRouter } from "react-router";
import { Helmet } from "react-helmet";
import { processHtml } from "./html-processor";
import App from "./../components/App/App";
import Error from "./../components/Error/Error";
import stringifyObject from "stringify-object";

const BaseUrlRelative: string = "/consultations";

// Returns a script tag with the stringified data for loading on the client
const getPreloadedDataHtml = (data): string => {
	let scriptTag: string = `<script>window.__PRELOADED__=${JSON.stringify(data)};</script>`;

	// Wrap new lines in dev mode so it's easier to scan over html source for debugging purposes
	if (process.env.NODE_ENV === "development") {
		scriptTag = `\r\n\r\n${scriptTag}\r\n\r\n`;
	}

	return scriptTag;
};

// Returns a promise that resolves to an object containing the HTML to be rendered.
// The params contains properties e.g.
// { location: { protocol, slashes, auth, host, port, hostname, hash, search, query: {}, pathname, path, href },
// origin, url, baseUrl, absoluteUrl, domainTasks: { }, data: { originalHtml: "", ... }
// The `params.data` property contains properties set in `SupplyData` in Startup.cs.
export const serverRenderer = (params): Promise => {
	return new Promise((resolve, reject) => {
		// Context object that Routes can use to pass properties 'out'. Primarily used for status code.
		// eslint-disable-next-line
		// See https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/StaticRouter.md#context-object
		// E.g.:
		//  <Route render={({ staticContext }) => {
		//      if (staticContext) staticContext.status = 404;
		//      return null; }} />
		let staticContext = {
			preload: {
				data: {
					cookies: params.data.cookies,
					isAuthorised: params.data.isAuthorised,
					displayName: params.data.displayName,
					signInURL: params.data.signInURL,
					registerURL: params.data.registerURL,
					requestURL: params.data.requestURL,
				}, // Key value pairs of preloaded data sets
				loaders: [], // List of promises where we track preloading data
			},
			baseUrl: params.origin + BaseUrlRelative,
			// Base url is used for 'server' ajax requests so we can hit the .NET instance from the Node process
		};

		let app = (
				<StaticRouter basename={BaseUrlRelative} location={params.url} context={staticContext}>
					<App />
				</StaticRouter>);

		// First render: this trigger any data preloaders to fire
		let rootContent = renderToString(app),
			statusCode: number = 200,
			html: string;
		
		// Wait for all preloaders to have loaded before re-rendering the app
		Promise.all(staticContext.preload.loaders).then(() => {
			
			// Second render now that all the data preloaders have finished so we can render with data on the server
			rootContent = renderToString(app);

			const helmet = Helmet.renderStatic();
			const html = processHtml(params.data.originalHtml,
				{
					htmlAttributes: helmet.htmlAttributes.toString(),
					bodyAttributes: helmet.bodyAttributes.toString(),
					rootContent: rootContent,
					title: helmet.title.toString(),
					metas: helmet.meta.toString(),
					links: helmet.link.toString(),
					scripts: getPreloadedDataHtml(staticContext.preload.data) + helmet.script.toString(),
				});

			resolve({ html: html, statusCode: staticContext.status || 200 });
			
		}).catch((e) => {
			if (process.env.NODE_ENV === "production") {
				// In production, rejecting the promise shows a standard dotnet 500 server error page
				reject(e);
				return;
			}
			// In development show a nice YSOD to devs with the error message
			const error = <Error error={e} />;
			rootContent = renderToString(error);
			
			resolve({ html: rootContent, statusCode: staticContext.status || 500 });
			return;
			
		});
	});
};

// `createServerRenderer` is what the DotNetCore SpaServices requires for SSR
export default createServerRenderer(serverRenderer);
