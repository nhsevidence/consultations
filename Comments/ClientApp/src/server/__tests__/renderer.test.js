import { Helmet } from "react-helmet";
import { serverRenderer } from "../renderer";

// To avoid "Error: You may ony call rewind() on the server. Call peek() to read the current state."
Helmet.canUseDOM = false;

describe("Server renderer", () => {
	describe("serverRenderer", () => {

		it("rejects promise when app rendering fails in production", (done) => {
			process.env.NODE_ENV = "production";

			serverRenderer(null).catch((e) => {
				expect(e.message).toEqual("Cannot read property 'data' of null");
				done();
			});
		});

		it("resolves promise with error component and 500 status when app rendering fails in development", (done) => {
			process.env.NODE_ENV = "development";

			serverRenderer({ data: { viewModel: 99 } }).then((result) => {
				const expectedResult = {
					html: "<div data-reactroot=\"\">error</div>",
					statusCode: 500,
					globals: {
						viewModel: 99
					}
				};
				expect(result).toEqual(expectedResult);
				done();
			});
		});


		it("returns a promise that resolves with correct html", () => {
			var result = serverRenderer({ url: "/",
				data: { originalHtml: "<div>test-html</div>" } });
			return expect(result).resolves.toHaveProperty("html", "<div>test-html</div>");
		});
	});
});
