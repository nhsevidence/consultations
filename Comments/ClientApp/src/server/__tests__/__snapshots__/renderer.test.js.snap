// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server renderer serverRenderer resolves promise with error component and 500 status when app rendering fails in development 1`] = `
"<div class=\\"container\\" data-reactroot=\\"\\"><div class=\\"alert\\"><h2 class=\\"page-header__heading mt--0\\"><span class=\\"icon icon--warning\\" aria-hidden=\\"true\\"></span> <!-- -->TypeError</h2><p class=\\"page-header__lead\\">Cannot read property &#x27;replace&#x27; of undefined</p><p>TypeError: Cannot read property &#x27;replace&#x27; of undefined
    at Object.&lt;anonymous&gt;.exports.replaceRelativePaths.html (C:\\\\_src\\\\consultations\\\\Comments\\\\ClientApp\\\\src\\\\server\\\\html-processor.js:60:14)
    at Object.&lt;anonymous&gt;.exports.processHtml (C:\\\\_src\\\\consultations\\\\Comments\\\\ClientApp\\\\src\\\\server\\\\html-processor.js:70:9)
    at Promise.all.then (C:\\\\_src\\\\consultations\\\\Comments\\\\ClientApp\\\\src\\\\server\\\\renderer.js:75:42)
    at &lt;anonymous&gt;
    at process._tickCallback (internal/process/next_tick.js:188:7)</p></div></div>"
`;
